aws_xray_sdk/core/serverless_context.py

```diff
@@ -0,0 +1,131 @@
+        parent_facade_segment = self.__get_facade_entity()  # type: FacadeSegment
+        mimic_segment = MimicSegment(parent_facade_segment, segment)
+        parent_facade_segment.add_subsegment(mimic_segment)
+        Context.put_segment(self, mimic_segment)
+
+    def end_segment(self, end_time=None):
+        """
+        Close the MimicSegment
+        """
+        # Close the last mimic segment opened then remove it from our facade segment.
+        mimic_segment = self.get_trace_entity()
+        Context.end_segment(self, end_time)
+        if type(mimic_segment) == MimicSegment:
+            # The facade segment can only hold mimic segments.
+            facade_segment = self.__get_facade_entity()
+            facade_segment.remove_subsegment(mimic_segment)
+
+    def put_subsegment(self, subsegment):
+        """
+        Appends the subsegment as a subsegment of either the mimic segment or
+        another subsegment if they are the last opened entity.
+        :param subsegment: The subsegment to to be added as a subsegment.
+        """
+        Context.put_subsegment(self, subsegment)
+
+    def end_subsegment(self, end_time=None):
+        """
+        End the current subsegment. In our case, subsegments
+        will either be a subsegment of a mimic segment or another
+        subsegment.
+        :param int end_time: epoch in seconds. If not specified the current
+            system time will be used.
+        :return: True on success, false if no parent mimic segment/subsegment is found.
+        """
+        return Context.end_subsegment(self, end_time)
+
+    def __get_facade_entity(self):
+        """
+        Retrieves the Facade segment from thread local. This facade segment should always be present
+        because it was generated by the Lambda Container.
+        :return: FacadeSegment
+        """
+        self._refresh_context()
+        facade_segment = self._local.segment  # type: FacadeSegment
+        return facade_segment
+
+    def get_trace_entity(self):
+        """
+        Return the latest entity added. In this case, it'll either be a Mimic Segment or
+        a subsegment. Facade Segments are never returned.
+        If no mimic segments or subsegments were ever passed in, throw the default
+        context missing error.
+        :return: Entity
+        """
+        # Call to Context.get_trace_entity() returns the latest mimic segment/subsegment if they exist.
+        # Otherwise, returns None through the following way:
+        # No mimic segment/subsegment exists so Context calls LambdaContext's handle_context_missing().
+        # By default, Lambda's method returns no-op, so it will return None to ServerlessContext.
+        # Take that None as an indication to return the rightful handle_context_missing(), otherwise
+        # return the entity.
+        entity = Context.get_trace_entity(self)
+        if entity is None:
```

Comment:
Interestingly lambda facade segment will always be present even if somewhere in the web app doesn't propagate context properly. Should we attach whatever we have to the facade segment and throw a warning? There is no meaning to force fail user's application when there IS context available.

Reply:
The facade segment is stored in threadlocal's 'segment' field, not the 'entities' list. In the lambda context implementation of get_trace_entity(), if no entities exist in the 'entities' list, then it returns threadlocal.segment, which is the facade segment. In this design, the FacadeSegment is never actually revealed to the higher ups and can't be accessed outside of the context itself. This design gives the customer the abstraction that they can BeginSegments, mutate it, and append subsegments onto it, but only if they have created a segment (which becomes a mimic segment)

This handle_context_missing() would be in the case where the higher-level calls didn't generate a segment. When a segment is not generated, we want to give the abstraction of what would normally happen in the grandparent class "Context", so in this case, throw an exception since the MimicSegment was never generated.